cmake/utils.cmake

```diff
@@ -138,3 +138,32 @@ function(set_hardening_flags)
         endif()
     endif()
 endfunction()
+
+#
+# Sets the `MACRO_PREFIX_MAP_FLAG` and `DEBUG_PREFIX_MAP_FLAGS` used for `-fmacro-prefix-map` and `-fdebug-prefix-map`
+# with `-Wa,--debug-prefix-map` compiler flags respectively.
+#
+# Uses the `DEV_BUILD` option.
+function(set_prefix_map_flags)
+    # Assuming the C and C++ compiler IDs are equal.
+    # For clang `MATCHES` is used to match both Clang and AppleClang.
+    if ((CMAKE_C_COMPILER_ID MATCHES "Clang" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL "10") OR
+        (CMAKE_C_COMPILER_ID STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL "8"))
+        set(MACRO_PREFIX_MAP_FLAG "-fmacro-prefix-map=${CMAKE_SOURCE_DIR}=." PARENT_SCOPE)
+    endif()
+
+    # Assuming the C and C++ compiler IDs are equal.
+    # For clang `MATCHES` is used to match both Clang and AppleClang.
+    # Since this flag requires an extra command to tell the debugger where to find
+    # the source files, which is not convenient for developers, it is disabled when
+    # `DEV_BUILD` option is turned on.
+    if (NOT DEV_BUILD AND
+        ((CMAKE_C_COMPILER_ID MATCHES "Clang" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL "3.8") OR
+        (CMAKE_C_COMPILER_ID STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL "4.4.7")))
+        # The AppleClang assembler does not support this option.
+        if (NOT APPLE)
+            set(AS_DEBUG_PREFIX_MAP_FLAG "-Wa,--debug-prefix-map=${CMAKE_SOURCE_DIR}=.")
+        endif()
+        set(DEBUG_PREFIX_MAP_FLAGS "-fdebug-prefix-map=${CMAKE_SOURCE_DIR}=. ${AS_DEBUG_PREFIX_MAP_FLAG}" PARENT_SCOPE)
```

Comment:
There is no requirement to don't have (build) system paths in the executable or logs. It is not recommended to show them in outgoing messages. Let's strip it from them? This way we don't need to break `gdb.sh`, splitdebug installs, local debugging and likely other dev/user/admin flows that lean on usual paths in the debug information.

Reply:
We discussed this in person with @locker and @alyapunov: it looks like a reasonable default for production builds.

Reply:
With all my respect to the colleagues, I would like to hear arguments, not names. If it is reasonable, there should be a reason. Something like 'the alternative would be X, which has disadvantages Y and Z'.

I'll repeat my points.

* Is `tarabrt.sh` + `gdb.sh` flow becomes broken?
* Whether we decided to make user side debugging harder?

What exactly I should pass to `gdb` to make the new executable debuggable:

1. When building from sources.
2. When I installed a source package from a repository.

There is also a minor point regarding splitdebug (portage feature) -- I guess it'll not work too. But I can fix it in the ebuild.

Reply:
> Is tarabrt.sh + gdb.sh flow becomes broken?

Yes, it requires an extra command, `set substitute-path`.

> Whether we decided to make user side debugging harder?

What do you mean by 'user side debugging'?

> When I installed a source package from a repository.

What is a 'source package from a repository'? If I understand what you mean, I don't quite get how you debug it in the first place.

> When building from sources.

You can either turn on `DEV_BUILD` option or run the **GDB** `set substitute-path . <path to the root of tarantool source tree>`.

Reply:
> Something like 'the alternative would be X, which has disadvantages Y and Z'

The problem with absolute paths in debugging infromation is that they are a redundant build artefact for users who you use Tarantool in production: they are only needed by developers. 

The alternative would be to make the provided option have the opposite effect (and rename it to, let's say, `PROD_BUILD`: by default turned off, when turned on, all absolute paths in debugging information are replaced with relative ones.

Reply:
@locker @alyapunov RFC

Reply:
> Yes, it requires an extra command, `set substitute-path`.

Let's add it so, otherwise we'll definitely introduce a problem. And ensure that `gdb.sh` works as with old debuginfo format as well as with the new one.

> What do you mean by 'user side debugging'?

Debugging by a user (on a package installed from our repositories). At least ability to run `gdb` and show backtraces/variables. It is sometimes helpful, when a power user tells you details about a problem on its own. I would consider non-working gdb integration as a bug. It works for other packages, why not for tarantool?

> What is a 'source package from a repository'?

https://wiki.debian.org/Packaging/SourcePackage

Reply:
> The problem with absolute paths in debugging infromation is that they are a redundant build artefact for users who you use Tarantool in production: they are only needed by developers.

Ability to inspect an executable (including one installed from a package) using gdb without extra tarantool-specific magic is not redundant. Ability to use `tarabrt.sh` and `gdb.sh` is not redundant.

At least two bugs are introduced here.

Reply:
> At least ability to run gdb and show backtraces/variables

This doesn't break **GDB** backtraces.

Reply:
> Ability to inspect an executable (including one installed from a package) using gdb without extra tarantool-specific magic is not redundant. Ability to use tarabrt.sh and gdb.sh is not redundant.
> 
> At least two bugs are introduced here.

@Totktonada please address your concerns to security team in [TNT-625](https://jira.vk.team/browse/TNT-625).

Reply:
> AFAIC, this doesn't break GDB backtraces and variables, it only breaks stepping into source code.

Everything that need sources. I don't know how to inspect local variables in terms of C without sources. Anyway, gdb is much less useful without sources -- that's the point.

> @Totktonada please address your concerns to security team in [TNT-625](https://jira.vk.team/browse/TNT-625).

I asked questions regarding the patch from the author of the patch. It seems perfectly legal.

----

Let's step back to my initial questions:

> It is not recommended to show them in outgoing messages. Let's strip it from them?

> Something like 'the alternative would be X, which has disadvantages Y and Z'

>> The problem with absolute paths in debugging infromation is that they are a redundant build artefact for users who you use Tarantool in production: they are only needed by developers.

I asked to compare two approaches: your one and stripping path prefix in the error subsystem implementation. That's the alternative. Which disadvantages it has?

Regarding your answer: aside of usefulness of the 'redundant' information (see above) you incorrectly understood [CWE-497](https://cwe.mitre.org/data/definitions/497.html). There is no sense to protect system information from one, who able to read the executable file. The information shouldn't be shared with remote peers if it is unnecessary.

Reply:
> Anyway, gdb is much less useful without sources -- that's the point.

The point is, do you know of many users debugging tarantool and stepping into sources?

Reply:
> I asked questions regarding the patch from the author of the patch.

My patch is based on the interpetation our team has of the task.

Reply:
> I asked to compare two approaches: your one and stripping path prefix in the error subsystem implementation.

As you could read from the commit message, my patch addresses two separate concerns: 
1. replacing paths in diagnostics
2. replacing paths in debugging information

I can easily drop the second part, if we all agree on it.

Reply:
> The point is, do you know of many users debugging tarantool and stepping into sources?

I know some. Do you know many users, which definitely didn't? This direction of the discussion will not give us any decision: we all are not representative.

> My patch is based on the interpetation our team has of the task.

That's nice (it is good to have an agreement within a working group first) and, moreover, it is legal to completely ignore my opinion here (I guess). However if you want to convince me, 'we agreed' doesn't work.

> <...> my patch addresses two separate concerns:
>
> 1. <...>
> 2. replacing paths in debugging information.

My point is that there is no reason for the second item. TNT-625 seems based on CWE-497 (or a similar recommendation) and has sense so. The second point should also have some reason under the hood. As we see, there are downsides. If there are no advantages, it is not worthful to do.

> It seems you didn't quite understand what my patch does.
> <...>
> The first part doesn't affect debugging, and I can easily drop the second part, if we all agree on it.

I don't see how, but okay. I propose to revisit a goal first.

Reply:
I just looked at `tools/gdb.sh`, and I don't quite understand what we are arguing about:
https://github.com/tarantool/tarantool/blob/98fcd437d9bec3f8f35304aad7ab1fdc3822ff45/tools/gdb.sh#L105

AFAIS, `set substitute-path` is used anyways, so let's just patch it, and add `-ex "set substitute-path ${SUBPATH} ." \`?

Reply:
> As we see, there are downsides. If there are no advantages, it is not worthful to do.

As I said, the advantage is that we can make reproducible builds in the future.

Reply:
> I don't see how, but okay. I propose to revisit a goal first.

[`-fmacro-prefix-map`](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#Preprocessor-Options) flag is responsible for the first part and [`-fdebug-prefix-map`](https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html#Debugging-Options) and [`--debug-prefix-map`](https://linux.die.net/man/1/as) flags are reponsible for the second part: please visit the **GCC** and **GAS** documentation to get an idea of how they work.

Reply:
My 2c:
- Full paths (including the build directory) in a binary delivered to the user make no sense. Neither in error messages nor in debug info. They should be removed if possible.
- We should check that stripping the paths during build doesn't break debug info in our deb/rpm packages.
- Breaking our internal scripts is fine as long as we fix them.
- Are there any other projects that strip file paths on build?

Reply:
> Current implementation has its own price (broken integration with gdb) and there is no explicit plans regarding reproducible builds. Does it mean that in result we'll broke one thing and will not offer anything to a user instead?

From what I discussed with colleagues yesterday, everone say people use `set substitute-path` anyways and none really relies on the fact that source files mapping is correct. 

I'm kinda sick of this argument, I suggest we move it to server team chat.

Reply:
Actually, AFAIU from the regular expression here, `.` also matches it, and we don't break `tool/gdb.sh`...

https://github.com/tarantool/tarantool/blob/761574e33c514aa36f17baf426b06f67ba5c03c6/tools/gdb.sh#L96-L100

Reply:
I setup a meeting. Anyway, there is no motivation to make this change. It doesn't give anything: neither reproducibility, nor security. I don't understand why you don't just do it for youself, nobody needs it.

> everone says people use `set substitute-path` anyways and noone really relies on the fact that source files mapping is correct.

It seems, all the discussion is about 'one person said', 'another person said' and so on. Okay. [Someone](https://stackoverflow.com/questions/67191647/using-ffile-prefix-map-breaks-debugging#comment118797400_67197241) said that it does not work. Let's say me that someone else said that it'll work. Of course, don't check it.

Reply:
Tested: remapping from relative path to sources path in `tools/gdb.sh` works correctly:
<img width="314" alt="Screen Shot 2022-10-20 at 13 42 26" src="https://user-images.githubusercontent.com/32464072/196927762-3d8adc74-c4a7-460f-a3f1-6a1fba45a06b.png">

Reply:
Okay so. We discussed it (me, @CuriousGeorgiy, @locker, @kyukhin) and agreed to take the risk of the subtle changes and proceed with stripping of the build directory from the debug info.